### PR TITLE
Add Live Demo Script

### PR DESCRIPTION
**Pull request by lichtefeld**
Currently a WIP dive demo script. Preprocessing integration is still needed before this can be tested fully.

### CODE REVIEW

**Review by spigo900 — COMMENTED** [review]
Not a formal review, just noting a few things I noticed before I forget.

File: adam/curriculum/curriculum_from_files.py [hunk 1]
```diff
@@ -98,59 +99,67 @@ def phase3_load_from_disk(  # pylint: disable=unused-argument
...
-                frames=[
-                    VisualPerceptionFrame.from_yaml(
-                        situation_dir / feature_file, color_is_rgb=color_is_rgb
-                    )
-                    for feature_file in feature_yamls
-                ],
-                action_features=action_features,
-            )
-        instances.append((situation, language, perception))  # type: ignore
+        instances.append(phase3_process_scene_dir(situation_dir, color_is_rgb=color_is_rgb, curriculum_type=curriculum_type))  # type: ignore
```
Comment (spigo900) [inline adam/curriculum/curriculum_from_files.py:102]: I think we can drop the type ignore here. mypy doesn't complain when I run locally with `mypy==0.942`.

File: adam/experiment/live_demo.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+from adam.learner.integrated_learner import SimulatedIntegratedTemplateLearner
+from adam.paths import (
+    TESTING_CURRICULUM_DIR,
+    LEARNERS_DIR,
+    PREPROCESSING_DIR,
+    TRAINING_CURRICULUM_DIR,
+)
+
+MENU_STR = "ADAM Demo UI Valid Commands: (d)ecode, (e)xit, (r)eset curriculum: "
+DECODE_STR = "Filepath to the image to decode: "
```
Comment (spigo900) [inline adam/experiment/live_demo.py:26]: For clarity I'd call these `MENU_PROMPT` and `DECODE_PROMPT` instead. `STR` is redundant since we're assigning literal values and `PROMPT` is more specific about how they're used.

File: adam/experiment/live_demo.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+                    linguistic_description,
+                    perception_frame,
+                    descriptions_from_learner,
+                )
+                observer.report()
+
+                print(f"Demo scene {curriculum.situation_number} has been processed.")
+            # We purposely catch broad exceptions here because we don't want the interactive session to end
+            # If an exception occurs in processing a given file
+            except Exception as e:  # pylint: disable=broad-except
```
Comment (spigo900) [inline adam/experiment/live_demo.py:307]: Not sure where to put this, but I ran into a problem while debugging before the demo -- is this still a thing or is it handled? The problem was that if preprocessing fails, then the demo curriculum path would be left in an inconsistent state where we have say 1 curriculum dir but no `info.yaml`, or 3 and the info file says 2.

Before we just crashed when this happened which made the issue more dire. When I restarted the script it wouldn't be able to process further images, because it would try to write to a directory that shouldn't exist but does (e.g. `situation_0`). I had to delete the directory and restart, because `(r)eset` didn't work at the time (the directory was no longer empty).

I'm not sure if this is still a problem now that we're catching exceptions. It's possible that because we're maintaining the curriculum state things might be okay. The same problem might come up though if you quit and restart after preprocessing crashes?

File: adam/experiment/live_demo.py [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+        )
+        rng = np.random.default_rng(self.color_seed)
+        segment_colors = np.zeros((len(masks) + 1, 3), dtype=np.uint8)
+        segment_colors[1:, :] = rng.uniform(0, 256, size=(len(masks), 3)).astype(np.uint8)
+        # Convert grayscale to BGR
+        # bgr_segmentation_ids = cv2.cvtColor(segmentation_ids, cv2.COLOR_GRAY2BGR)
+        bgr_segmentation_ids = np.take(segment_colors, indices=segmentation_ids, axis=0)
+        print("Images processed.")
+
+        # ETA: forgot to include saving 🙃
```
Comment (spigo900) [inline adam/experiment/live_demo.py:135]: Think we can delete this now:
```suggestion
```

File: adam/experiment/live_demo.py [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+                axis=0,
+            )
+            if masks.size > 0
+            else np.zeros(img.shape[:2], dtype=np.uint32)
+        )
+        rng = np.random.default_rng(self.color_seed)
+        segment_colors = np.zeros((len(masks) + 1, 3), dtype=np.uint8)
+        segment_colors[1:, :] = rng.uniform(0, 256, size=(len(masks), 3)).astype(np.uint8)
+        # Convert grayscale to BGR
+        # bgr_segmentation_ids = cv2.cvtColor(segmentation_ids, cv2.COLOR_GRAY2BGR)
```
Comment (spigo900) [inline adam/experiment/live_demo.py:131]: We don't need this anymore:
```suggestion
```

File: adam/experiment/live_demo.py [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+        img = cv2.imread(str(image_path))
+        img2 = None
+        try:
+            img2 = cv2.cvtColor(img, cv2.COLOR_RGBA2RGB)
+            print("Image is RGBA")
+        except ValueError:
+            print("Image is RGB")
+            img2 = img
+        finally:
+            cv2.imwrite(str(situation_img_path), img2)
```
Comment (spigo900) [inline adam/experiment/live_demo.py:88]: (No change needed) It bothers me that I feel like we could do something smarter here... but we don't need to, so 🤷‍♂️.

File: adam/experiment/live_demo.py [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+        resp = requests.post(api, data=payload, headers=headers)
+
+        data = resp.json()
+
+        print("Generating Segmentation Mask")
+        # 3D array: (N, W, H) where N is number of masks
+        masks = np.asarray(data["masks"], dtype=np.uint8)
+        print(f"Got {len(masks)} masks.")
+
+        # Assert masks don't overlap
```
Comment (spigo900) [inline adam/experiment/live_demo.py:107]:
```suggestion
        # Check for and warn about overlapping masks. We warn because this is an edge case
        # where we don't have a good solution. Currently we deal with it using an arbitrary hack
        # of "prefer the later mask", which might give weird results sometimes.
```

File: adam/experiment/live_demo.py [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+        for i in range(len(masks)):  # pylint: disable=consider-using-enumerate
+            for j in range(i + 1, len(masks)):  # pylint: disable=consider-using-enumerate
+                if np.any(masks[i] & masks[j]):
+                    print(
+                        f"Warning: Masks {i} and {j} overlap. Preferring later mask {j}."
+                    )
+
+        # Compute segmentation IDs
```
Comment (spigo900) [inline adam/experiment/live_demo.py:115]: I want to explain more about the details of how this is computed because it's pretty gnarly, but I can't think of a good comment there, so maybe it is more useful to simply expand on what this expression computes:

```suggestion
        # Compute pixelwise segmentation/mask IDs
        #
        # That is, an input-image-sized array (shape (W, H)) where each entry is either 0 (meaning
        # not in any mask) or a positive integer identifying which of the `N` output masks this pixel
        # belongs to. When a pixel lies in the overlap of several masks, we arbitrarily take the
        # highest-numbered mask.
```

File: adam/experiment/live_demo.py [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+                masks
+                * (1 + np.arange(masks.shape[0], dtype=masks.dtype))[
+                    :, np.newaxis, np.newaxis
+                ],
+                axis=0,
+            )
+            if masks.size > 0
+            else np.zeros(img.shape[:2], dtype=np.uint32)
+        )
+        rng = np.random.default_rng(self.color_seed)
```
Comment (spigo900) [inline adam/experiment/live_demo.py:127]: Guide comment on what we're doing with the IDs?

```suggestion
        # Create segmentation mask image by assigning random colors to each mask
        rng = np.random.default_rng(self.color_seed)
```

**Review by spigo900 — CHANGES_REQUESTED** [review]
Commenting with a bunch of small cleanup changes. Only four functional issues I think -- one to fix GNN training, one about calculating GNN accuracy, one about resetting the curriculum, one question about what happens when we crash during preprocessing.

One other thing: we should add requirements somewhere for `opencv-python`, `requests`, and `types-requests`. Probably add `types-requests>=2.28.8` to our dev requirements -- I ran `make mypy` and it complained about lacking `types-requests`.

File: adam/experiment/live_demo.py (hunk 3 above)
Comment (lichtefeld) [reply adam/experiment/live_demo.py:307]: The goal was that catching exceptions should keep the curriculum manager from going out-of-sync with disk so when we go to process a new image everything works fine. I didn't thoroughly test this but it did appear to work in theory.

File: adam/experiment/live_demo.py (hunk 6 above)
Comment (lichtefeld) [reply adam/experiment/live_demo.py:88]: Agreed, happy to call this "live demo code" and not come up with a better solution unless needed.

File: adam_preprocessing/shape_stroke_graph_inference.py [hunk 10]
```diff
@@ -106,32 +112,45 @@ def main():
...
+        )
+    else:
+        # if we know there's only one input, we know there should only be one output, so we CAN give
+        # a placeholder for now
+        if args.dir_num is not None:
+            logging.warning(
+                "Can't give reasonable output for empty input. Outputting default decode of 0 (%s)",
+                STRING_OBJECT_LABELS[0],
+            )
+            outputs = torch.zeros((1, 1))
```
[the remark this answers is not in the document]
Comment (lichtefeld) [reply adam_preprocessing/shape_stroke_graph_inference.py:127]: If we have no strokes maybe the reasonable solution is to just not provide any output and call it a miss from processing error? Then we'd just have output with no language from ADAM but that's "ok" if scene processing has failed

[the remark this answers is not in the document]
Comment (spigo900) [reply adam_preprocessing/shape_stroke_graph_inference.py:127]: That works for me, and I think that works fine for the demo, but I am slightly worried about a crash in batch mode. I don't recall how well ADAM handles it when it's missing a feature file/the feature file is missing any actual features.

[the remark this answers is not in the document]
Comment (lichtefeld) [reply adam_preprocessing/shape_stroke_graph_inference.py:127]: My guess is that it'll just wreck the pattern but I'm not sure.

File: adam/experiment/live_demo.py [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+
+        if initialize_after_reset:
+            self.initialize_curriculum()
+
+    def preprocess(
+        self, image_path: Path, host: str, stroke_model_path: Path, stroke_python: Path
+    ) -> Path:
+        """Preprocess an image with stroke extraction to configure the curriculum scene."""
+        situation_dir = self.curriculum_path / f"situation_{self.situation_number}"
+        situation_dir.mkdir(parents=True)
```
[the remark this answers is not in the document]
Comment (lichtefeld) [reply adam/experiment/live_demo.py:76]: hmm, if I just add `exists_ok=True` than that at least solves the "we can't create the same directory twice" bug. I think every other operation in preprocessing will just overwrite files which I believe is fine given the directory is in a corrupted state. Does that solution sound fine to you?

[the remark this answers is not in the document]
Comment (spigo900) [reply adam/experiment/live_demo.py:76]: Yes, that's fine to me.

File: adam_preprocessing/shape_stroke_graph_inference.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (spigo900) [reply adam_preprocessing/shape_stroke_graph_inference.py:127]: Based on `phase3_load_from_disk()` it looks like when loading we should end up with no feature files when we glob for them, so `len(feature_yamls) != 1`, so we will try to load `action.yaml` which doesn't exist and will crash.

[the remark this answers is not in the document]
Comment (lichtefeld) [reply adam_preprocessing/shape_stroke_graph_inference.py:127]: Ah, ok, I'd want to update this so we still provide a `feature.yaml` file just the stroke features would be empty. That way we import an empty file and ADAM doesn't crash.

[the remark this answers is not in the document]
Comment (spigo900) [reply adam_preprocessing/shape_stroke_graph_inference.py:127]: I think that makes sense.

**Review by spigo900 — APPROVED** [review]
Looks good to me, two small things I'll comment on. (ETA: two replies to threads, that is -- which should be up.)

File: adam_preprocessing/shape_stroke_graph_inference.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (spigo900) [reply adam_preprocessing/shape_stroke_graph_inference.py]: I think you want to drop the None check on [line 138](https://github.com/isi-vista/adam/pull/1159/files#diff-66e64562fb1b22bb6e3daf6ac5d35aa750717246813016ea80e206d8cb193ce8R138) so that we produce blank/placeholder output when there are no strokes to use to identify the input. But note this only produces placeholder output within the context of the demo.

We have to extend the code if we want to do this for a general curriculum, i.e. not just within the demo. That would probably involve changes to the stroke extraction code too. That shouldn't be hard but it's involved enough I think it's not worth tackling in this PR.

File: adam_preprocessing/shape_stroke_extraction.py [hunk 12]
```diff
@@ -537,21 +535,30 @@ def main():
...
     # copied and edited from phase3_load_from_disk() -- see adam.curriculum.curriculum_from_files
-    with open(
-        args.curriculum_path / "info.yaml", encoding="utf=8"
-    ) as curriculum_info_yaml:
-        curriculum_params = yaml.safe_load(curriculum_info_yaml)
-    logger.info("Input curriculum has %d dirs/situations.", curriculum_params["num_dirs"])
+    if args.dir_num:
+        logger.info(f"Processing directory number {args.dir_num}")
+
+    if not args.dir_num:
```
[the remark this answers is not in the document]
Comment (spigo900) [reply adam_preprocessing/shape_stroke_extraction.py:551]: Missed?

File: adam_preprocessing/shape_stroke_graph_inference.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (lichtefeld) [reply adam_preprocessing/shape_stroke_graph_inference.py]: Agreed with handling the more complex interaction in another PR. I've relocated the check for `outputs` being None to guard against the assert clause for now.

[the remark this answers is not in the document]
Comment (spigo900) [reply adam_preprocessing/shape_stroke_graph_inference.py]: Sounds good.
